File: app/src/main/java/com/beautifulquran/domain/OutputLatency.kt
package com.beautifulquran.domain

/**
 * Route-based output-latency presets for the karaoke clock.
 *
 * [HighlightEngine] is pure: it answers "which word at time *t*?". Bluetooth
 * (and similar) delay is a property of the *playback path*, not of the
 * segments — so the reader subtracts a small offset from
 * `player.positionMs` before [HighlightClock] / [HighlightEngine] see it.
 *
 * Android never exposes a reliable end-to-end "ms until the ear" number for
 * A2DP. These presets are intentionally coarse: good enough that ink lands
 * with the voice on typical headsets, without baking lag into timing data or
 * the engine. A later user nudge can sit on top if needed.
 */
object OutputLatency {

    /** Connected output kinds the monitor can report. Multiple may be present. */
    enum class OutputKind {
        /** Phone speaker, wired, USB — treat as near-zero extra lag. */
        LOCAL,
        /** Classic Bluetooth A2DP (and hearing-aid media). */
        BLUETOOTH_A2DP,
        /** LE Audio / BLE headset or speaker. */
        BLUETOOTH_LE,
    }

    /** The single route used for the preset table (highest-latency wins). */
    enum class Route {
        LOCAL,
        BLUETOOTH_A2DP,
        BLUETOOTH_LE,
    }

    const val LOCAL_MS = 0L
    /** Typical classic A2DP stack delay; devices vary ~100–300 ms. */
    const val A2DP_MS = 180L
    /** LE Audio is usually lower-latency than classic A2DP. */
    const val LE_MS = 80L

    /**
     * Pick one route from the set of currently available output kinds.
     * Prefer A2DP over LE over local so a connected BT headset is not
     * ignored when the built-in speaker is still listed as an output.
     */
    fun classify(kinds: Set<OutputKind>): Route = when {
        OutputKind.BLUETOOTH_A2DP in kinds -> Route.BLUETOOTH_A2DP
        OutputKind.BLUETOOTH_LE in kinds -> Route.BLUETOOTH_LE
        else -> Route.LOCAL
    }

    fun latencyMs(route: Route): Long = when (route) {
        Route.LOCAL -> LOCAL_MS
        Route.BLUETOOTH_A2DP -> A2DP_MS
        Route.BLUETOOTH_LE -> LE_MS
    }

    fun latencyMs(kinds: Set<OutputKind>): Long = latencyMs(classify(kinds))

    /**
     * Media-timeline position adjusted so the highlight tracks what the
     * listener *hears*, not the decoder playhead.
     */
    fun heardMs(mediaPositionMs: Long, latencyMs: Long): Long =
        (mediaPositionMs - latencyMs).coerceAtLeast(0L)

    /**
     * Karaoke query time fed to [HighlightEngine]: start from the heard
     * playhead, then advance by [leadMs] so word ink can run *ahead* of the
     * segment table (Ink Lab → Highlight lead). Default lead is 0.
     *
     * Net form (not sequential clamp) so lag and lead cancel cleanly:
     * `max(0, media − latency + lead)` once the lead is fully engaged.
     *
     * When [leadNotBeforeMs] is positive, encoded opening silence stays on the
     * heard clock. After the first word starts, lead **ramps** from 0 to
     * [leadMs] over the first [leadMs] of voiced audio so engagement is
     * continuous with silence. A hard `+lead` cliff at the gate is larger than
     * [HighlightClock]'s post-handoff settle step and freezes the clock through
     * word 1 — both words then light when settle ends on word 2.
     */
    fun highlightMs(
        mediaPositionMs: Long,
        latencyMs: Long,
        leadMs: Long = 0L,
        leadNotBeforeMs: Long = 0L,
    ): Long {
        val heardPositionMs = heardMs(mediaPositionMs, latencyMs)
        val lead = leadMs.coerceAtLeast(0L)
        val gate = leadNotBeforeMs.coerceAtLeast(0L)
        // No opening silence: full lead from the first sample (lab default path).
        if (gate <= 0L) {
            return (mediaPositionMs - latencyMs.coerceAtLeast(0L) + lead)
                .coerceAtLeast(0L)
        }
        if (heardPositionMs < gate) return heardPositionMs
        // Ramp lead after the gate so engagement is continuous with silence.
        val pastGate = heardPositionMs - gate
        val appliedLead = minOf(lead, pastGate)
        return (heardPositionMs + appliedLead).coerceAtLeast(0L)
    }
}

File: web/src/domain/OutputLatency.ts
/**
 * Route-based output-latency presets for the karaoke clock.
 * Port of Android `domain/OutputLatency.kt`.
 *
 * HighlightEngine stays pure: Bluetooth (and similar) delay is a property of
 * the playback path. The reader subtracts a small offset from media position
 * before HighlightClock / HighlightEngine see it.
 *
 * Web has no reliable “ms until the ear” API, so route classification stays
 * available for a future monitor; the reader currently uses LOCAL (0 ms) lag
 * and still applies word-ink highlight lead via [highlightMs].
 */

export enum OutputKind {
  LOCAL = 'LOCAL',
  BLUETOOTH_A2DP = 'BLUETOOTH_A2DP',
  BLUETOOTH_LE = 'BLUETOOTH_LE',
}

export enum OutputRoute {
  LOCAL = 'LOCAL',
  BLUETOOTH_A2DP = 'BLUETOOTH_A2DP',
  BLUETOOTH_LE = 'BLUETOOTH_LE',
}

export const LOCAL_MS = 0
/** Typical classic A2DP stack delay; devices vary ~100–300 ms. */
export const A2DP_MS = 180
/** LE Audio is usually lower-latency than classic A2DP. */
export const LE_MS = 80

/** Prefer A2DP over LE over local so a connected BT headset is not ignored. */
export function classify(kinds: ReadonlySet<OutputKind>): OutputRoute {
  if (kinds.has(OutputKind.BLUETOOTH_A2DP)) return OutputRoute.BLUETOOTH_A2DP
  if (kinds.has(OutputKind.BLUETOOTH_LE)) return OutputRoute.BLUETOOTH_LE
  return OutputRoute.LOCAL
}

export function latencyMsForRoute(route: OutputRoute): number {
  switch (route) {
    case OutputRoute.BLUETOOTH_A2DP:
      return A2DP_MS
    case OutputRoute.BLUETOOTH_LE:
      return LE_MS
    default:
      return LOCAL_MS
  }
}

export function latencyMs(kinds: ReadonlySet<OutputKind>): number {
  return latencyMsForRoute(classify(kinds))
}

/** Media-timeline position adjusted so the highlight tracks what is heard. */
export function heardMs(mediaPositionMs: number, latencyMsValue: number): number {
  return Math.max(0, Math.trunc(mediaPositionMs) - Math.max(0, Math.trunc(latencyMsValue)))
}

/**
 * Karaoke query time: heard playhead, then advance by [leadMs] so word ink can
 * run ahead of the segment table. Net form so lag and lead cancel cleanly.
 *
 * When [leadNotBeforeMs] is positive, encoded opening silence stays on the
 * heard clock; after the first word starts, lead ramps from 0 to [leadMs].
 */
export function highlightMs(
  mediaPositionMs: number,
  latencyMsValue: number,
  leadMs = 0,
  leadNotBeforeMs = 0,
): number {
  const heardPositionMs = heardMs(mediaPositionMs, latencyMsValue)
  const lead = Math.max(0, Math.trunc(leadMs))
  const gate = Math.max(0, Math.trunc(leadNotBeforeMs))
  if (gate <= 0) {
    return Math.max(
      0,
      Math.trunc(mediaPositionMs) - Math.max(0, Math.trunc(latencyMsValue)) + lead,
    )
  }
  if (heardPositionMs < gate) return heardPositionMs
  const pastGate = heardPositionMs - gate
  const appliedLead = Math.min(lead, pastGate)
  return Math.max(0, heardPositionMs + appliedLead)
}

export const OutputLatency = {
  OutputKind,
  Route: OutputRoute,
  LOCAL_MS,
  A2DP_MS,
  LE_MS,
  classify,
  latencyMsForRoute,
  latencyMs,
  heardMs,
  highlightMs,
}
